### PR TITLE
reworked regrid to eval usr grid, but 0 every where else

### PR DESCRIPTION
**Pull request by dtnguyen2**
# Note

This PR implements the requirements as written by @DougBurke, see:
[@DougBurke requirements](https://icxc.harvard.edu/pipe/sherpadev/2020/0063.html) for details.

# caveat
Kept the case where data is integrated but model is not, does that ever get used?

### CODE REVIEW

**Comment by DougBurke** [comment]
@dtnguyen2 - using the code at the end of this message (which is based on the tests in #724), I see the following behavior, where the second column is the "truth" and the third column the model evaluation after the regrid. As it's a step function (Box1D, going from 0 to 0.4 at x=3.1 and back to 0 at x=4.2) there's going to be some issues, but we can see it is rather drastic:

```
% python tst.py
Npts: 40 40 40
x=2.0  yorig=0.00  yregrid=0.00
x=2.1  yorig=0.00  yregrid=0.00
x=2.2  yorig=0.00  yregrid=0.00
x=2.3  yorig=0.00  yregrid=0.00
x=2.4  yorig=0.00  yregrid=0.00
x=2.5  yorig=0.00  yregrid=-0.00
x=2.6  yorig=0.00  yregrid=11746.16
x=2.7  yorig=0.00  yregrid=-413.32
x=2.8  yorig=0.00  yregrid=0.00
x=2.9  yorig=0.00  yregrid=3.05
x=3.0  yorig=0.00  yregrid=-0.56
x=3.1  yorig=0.40  yregrid=0.40
x=3.2  yorig=0.40  yregrid=0.42
x=3.3  yorig=0.40  yregrid=0.39
x=3.4  yorig=0.40  yregrid=0.40
x=3.5  yorig=0.40  yregrid=0.40
x=3.6  yorig=0.40  yregrid=0.40
x=3.7  yorig=0.40  yregrid=0.40
x=3.8  yorig=0.40  yregrid=0.40
x=3.9  yorig=0.40  yregrid=0.40
x=4.0  yorig=0.40  yregrid=0.40
x=4.1  yorig=0.40  yregrid=0.42
x=4.2  yorig=0.00  yregrid=0.16
x=4.3  yorig=0.00  yregrid=-0.00
x=4.4  yorig=0.00  yregrid=-0.21
x=4.5  yorig=0.00  yregrid=0.64
x=4.6  yorig=0.00  yregrid=-0.00
x=4.7  yorig=0.00  yregrid=-20.40
x=4.8  yorig=0.00  yregrid=219.05
x=4.9  yorig=0.00  yregrid=-0.00
x=5.0  yorig=0.00  yregrid=-172708.28
x=5.1  yorig=0.00  yregrid=51219437.27
x=5.2  yorig=0.00  yregrid=0.00
x=5.3  yorig=0.00  yregrid=0.00
x=5.4  yorig=0.00  yregrid=0.00
x=5.5  yorig=0.00  yregrid=0.00
x=5.6  yorig=0.00  yregrid=0.00
x=5.7  yorig=0.00  yregrid=0.00
x=5.8  yorig=0.00  yregrid=0.00
x=5.9  yorig=0.00  yregrid=0.00
```

where

```
%  cat tst.py
import numpy as np
from sherpa.models.basic import Box1D

xgrid = np.arange(2, 6, 0.1)

requested = np.arange(2.5, 5.1, 0.075)

mdl = Box1D()
mdl.xlow = 3.1
mdl.xhi = 4.2
mdl.ampl = 0.4

yorig = mdl(xgrid)
yregrid = mdl.regrid(requested)(xgrid)

print("Npts: {} {} {}".format(xgrid.size, yorig.size, yregrid.size))
for x, y1, y2 in zip(xgrid, yorig, yregrid):
    print("x={:3.1f}  yorig={:4.2f}  yregrid={:4.2f}".format(x, y1, y2))
```

**Comment by dtnguyen2** [comment]
That is because you are using the default interpolation method, ```neville```.   I ran your script but used ```interpolate``` instead of ```neville``` and I get:
```
Npts: 40 40 40
x=2.0  yorig=0.00  yregrid=0.00
x=2.1  yorig=0.00  yregrid=0.00
x=2.2  yorig=0.00  yregrid=0.00
x=2.3  yorig=0.00  yregrid=0.00
x=2.4  yorig=0.00  yregrid=0.00
x=2.5  yorig=0.00  yregrid=0.00
x=2.6  yorig=0.00  yregrid=0.00
x=2.7  yorig=0.00  yregrid=0.00
x=2.8  yorig=0.00  yregrid=0.00
x=2.9  yorig=0.00  yregrid=0.00
x=3.0  yorig=0.00  yregrid=0.00
x=3.1  yorig=0.40  yregrid=0.40
x=3.2  yorig=0.40  yregrid=0.40
x=3.3  yorig=0.40  yregrid=0.40
x=3.4  yorig=0.40  yregrid=0.40
x=3.5  yorig=0.40  yregrid=0.40
x=3.6  yorig=0.40  yregrid=0.40
x=3.7  yorig=0.40  yregrid=0.40
x=3.8  yorig=0.40  yregrid=0.40
x=3.9  yorig=0.40  yregrid=0.40
x=4.0  yorig=0.40  yregrid=0.40
x=4.1  yorig=0.40  yregrid=0.40
x=4.2  yorig=0.00  yregrid=0.13
x=4.3  yorig=0.00  yregrid=0.00
x=4.4  yorig=0.00  yregrid=0.00
x=4.5  yorig=0.00  yregrid=0.00
x=4.6  yorig=0.00  yregrid=0.00
x=4.7  yorig=0.00  yregrid=0.00
x=4.8  yorig=0.00  yregrid=0.00
x=4.9  yorig=0.00  yregrid=0.00
x=5.0  yorig=0.00  yregrid=0.00
x=5.1  yorig=0.00  yregrid=0.00
x=5.2  yorig=0.00  yregrid=0.00
x=5.3  yorig=0.00  yregrid=0.00
x=5.4  yorig=0.00  yregrid=0.00
x=5.5  yorig=0.00  yregrid=0.00
x=5.6  yorig=0.00  yregrid=0.00
x=5.7  yorig=0.00  yregrid=0.00
x=5.8  yorig=0.00  yregrid=0.00
x=5.9  yorig=0.00  yregrid=0.00
```

**Comment by DougBurke** [comment]
How do I change the interpolator then? I don't think (but am hoping to be proved wrong) that I can do it with the code sample I used, since there appears to be no way to change the `method` of the underlying `ModelDomainRegridder1D` object (which I think is where you change the interpolation method).

**Comment by dtnguyen2** [comment]
It's true that I hardwired the interpolation method.  I will take a look and see

**Comment by dtnguyen2** [comment]
Make the following changes:
```
from sherpa.utils import interpolate

yorig = mdl(xgrid)
regrid_model = mdl.regrid(requested)
regrid_model.wrapper.method = interpolate
yregrid = regrid_model(xgrid)
```
then you should be good to go.

**Comment by anetasie** [comment]
@dtnguyen2 Do we need to make `interpolate` a default method for use in regrid? or there are some other cases when `neville` would be a better choice?

**Comment by dtnguyen2** [comment]
The following comment 
```
        # The tests show that neville (for simple interpolation-style
        # analysis) is much-more accurate than linear_interp, so use
        # that. If the user cares more about speed than accuracy
        # then they can switch to sherpa.utils.linear_interp.
        # Note that I have not tested the speed, so I am just assuming
        # that linear_interp is faster than neville.
```
was taken from the constructor of the ```class ModelDomainRegridder1D```.  The default will probably be neither of the two, working on getting something better.

**Comment by dtnguyen2** [comment]
While working on finding a _better_ interpolation algorithm for the regrid work, I realized that the
case where data is integrated but the integrate flag is ```False``` (does such a case really exist?) I realized that the current code will not set the function values to zero on grid point not requested by the user:
```
            else:
                # The integrate flag is set to false, so just evaluate the model
                # and then interpolate using the grids midpoints.
                y = modelfunc(pars, eval_space.midpoint_grid, **kwargs)
                return interpolate(data_space.midpoint_grid,
                                   eval_space.midpoint_grid, y,
                                   function=self.method)
        else:
            return self.eval_non_integrated(pars, modelfunc,
                                            data_space.midpoint_grid,
                                            eval_space.midpoint_grid,
                                            **kwargs)
```

Will probably have to modify the call the else case above to call ```self.eval_non_integratred``` method as in the non-integrated case.

**Comment by dtnguyen2** [comment]
Once I decide on the interpolate method for the regrid code, do we want to make the interpolation algorithm to be used as a runtime option?  For example:
```
    model.regrid(somexarrays, interp=neville)
```

**Comment by dtnguyen2** [comment]
Pushed code to enable runtime change of interpolation method.  The default interpolation method uses the Akima algorithm.  The Akima interpolation requires at least four data points which should not be an issue for the 1-D regrid method

**Comment by DougBurke** [comment]
@dtnguyen2 - my current thinking is that for the "1d integrated" case (so Data1DInt/DataPHA) the default "interpolation" should just be done by sherpa.utils.rebin since I assume (but have not researched) that it is better at conserving flux than the other methods. Of course, `rebin` has a different interface than `neville` (or `akima`) which might be problematic.

Does any of this make sense?

**Comment by dtnguyen2** [comment]
Never mind for now that I accidentally left two print statements in the code, but the integrated data does call ```rebin```:
```
       if data_space.is_integrated:
            if self.integrate:
                # This should be the most common case
                y = modelfunc(pars, eval_space.grid[0], eval_space.grid[1],
                              **kwargs)
                return rebin(y, eval_space.grid[0], eval_space.grid[1],
                             data_space.grid[0], data_space.grid[1])
            else:
                # The integrate flag is set to false, so just evaluate the model
                # and then interpolate using the grids midpoints.
                return self.eval_non_integrated(pars, modelfunc,
                                                data_space.midpoint_grid,
                                                eval_space.midpoint_grid,
                                                **kwargs)
```

For regriding, the interpolation algorithm is only used for **non**-integrated data

File: sherpa/utils/akima.py [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        ids = numpy.nonzero(f12 > 1e-12 * numpy.max(f12))[0]
+        b = m1[1:n + 1]
+
+        b[ids] = (f1[ids] * m1[ids + 1] + f2[ids] * m1[ids + 2]) / f12[ids]
+        c = (3.0 * m - 2.0 * b[0:n - 1] - b[1:n]) / dx
+        d = (b[0:n - 1] + b[1:n] - 2.0 * m) / dx ** 2
+        return b, c, d, n, x, y
+
+
+if '__main__' == __name__:
```
Comment (DougBurke) [inline sherpa/utils/akima.py]: Can we have this moved into a test, so it's always run? Or at least some sort of a test suite for the akima code? It is run in at least one of the tests added in this PR, but isn't really "unit tested".

**Comment by dtnguyen2** [comment]
@DougBurke, did as you asked but got an error when I tried to push it:

```
(sherpa) [dtn@devel12 sherpa]$ git push myfork regrid_eval_only_on_usr_grid 
Username for 'https://github.com': dtnguyen2
Password for 'https://dtnguyen2@github.com': 
To https://github.com/dtnguyen2/sherpa
 ! [rejected]          regrid_eval_only_on_usr_grid -> regrid_eval_only_on_usr_grid (fetch first)
error: failed to push some refs to 'https://github.com/dtnguyen2/sherpa'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

**Comment by DougBurke** [comment]
@dtnguyen2 - it looks like your branch on disk hasn't been updated to reflect the recent changes you pushed. What does

```
% git log --format=oneline | head
```

return (and does it match the commits made to this branch on Github)?

**Comment by dtnguyen2** [comment]
```
[dtn@devel12 sherpa]$ git log --format=oneline | head
5ee6ce6c18440124b9df0cc52868ef32deb25382 added a test for akima
fc91a738c5f8dfc54619bb1d4ef8fac8ca8f3c79 oops, accidentally left a couple of print statements
ed18b9bec9978ae0124f91535949ba085a5ee723 add optional 'interp' algorithm to the 1-D regrid method
1201b35335f30bd18a4180359a8803dd9513f08c reworked regrid to eval usr grid, but 0 every where else
922b2e4b90050a01f6557f83f184a2a6930f6c83 Merge #734 (DougBurke) - Change datastack query_by_header_keyword to not error if keyword is missing
3002c71c9364288ff40084b3463e8add939fbb71 Merge pull request #739 from Marie-Terrell/post-4120-updates
0932df9ab8c001c6fc452ef0bfb25bbf9cc333e2 Add the 4.12.0 Zenodo DOI to the README
9bdfd12b56fe85fe7a5d95d914e28b8d6efbfa3d Merge pull request #738 from sherpa/release/4.12.0
dcce768ec95bdb19ae95abd984e3afb3762a1ef0 RELEASE_NOTES
f559c89aabab76eb6d4d815908a5fb3607bd970e RELEASE_NOTES
```

**Comment by DougBurke** [comment]
You're missing commit e68febe aka "Merge branch 'master' into regrid_eval_only_on_usr_grid" from your repository, which is why git is saying you can't update.

I suggest you rebase your branch onto the masgter branch and then force-push it. This will get you the change from e68febe. So, something like

```
% git remote -v
```

to check if you have https://github.com/sherpa/sherpa available (probably with a name of `origin`). If not then you can add it

```
% git remote add upstream https://github.com/sherpa/sherpa
```

Then ensure it us up to date:

% git fetch upstream
```

(replacing `upstream` by whatever name `git remote` gave if you didn't have to use the `add` command).

Then rebase your branch (again replacing `upstream` if necessary)

```
% git checkout regrid_eval_only_on_usr_grid
% git rebase upstream/master
```

and then if this has no problems, 

```
% git push myfork regrid_eval_only_on_usr_grid  --force
```

**Comment by dtnguyen2** [comment]
@DougBurke, that seems to work.  Thanks
```
[dtn@devel12 sherpa]$ git push myfork regrid_eval_only_on_usr_grid --force
Username for 'https://github.com': dtnguyen2
Password for 'https://dtnguyen2@github.com': 
Enumerating objects: 48, done.
Counting objects: 100% (48/48), done.
Delta compression using up to 20 threads.
Compressing objects: 100% (34/34), done.
Writing objects: 100% (34/34), 8.18 KiB | 1.36 MiB/s, done.
Total 34 (delta 28), reused 0 (delta 0)
remote: Resolving deltas: 100% (28/28), completed with 13 local objects.
To https://github.com/dtnguyen2/sherpa
 + e68febe6...5ee6ce6c regrid_eval_only_on_usr_grid -> regrid_eval_only_on_usr_grid (forced update)
```

**Comment by DougBurke** [comment]
@dtnguyen2 - you must have missed a step out as you hadn't rebased onto the current master branch. I've just done that for you and pushed the update to your branch.

File: sherpa/utils/tests/test_akima.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License along
+#  with this program; if not, write to the Free Software Foundation, Inc.,
+#  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+
+import numpy
+
```
Comment (DougBurke) [inline sherpa/utils/tests/test_akima.py]: import pytest

File: sherpa/utils/tests/test_akima.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    x = numpy.sort(numpy.random.random(10) * 10)
+    y = numpy.random.normal(0.0, 0.1, size=len(x))
+    myakima = akima.Akima(x, y)
+
+    num = 3
+    for ii in numpy.linspace(0.01, 0.05, num):
+        xx = numpy.arange(x[0], x[-1], 0.05)
+        z = akima.interpolate(x, y, xx)
+        akima_z = myakima(xx)
+        if numpy.allclose(akima_z, z) is False:
```
Comment (DougBurke) [inline sherpa/utils/tests/test_akima.py]: This `if ...` section can be replaced by

    assert akima_z == pytest.approx(z)

which will report on differences in the input arrays if they are different.

**Comment by DougBurke** [comment]
ta for the test addition and update

**Comment by dtnguyen2** [comment]
@DougBurke, I had asked the following question:

> Kept the case where data is integrated but model is not, does that ever get used?

when I released the PR. So I assume that such a case does exist, if that is the case then we don't have too many tests because by commenting out lines [664, 670] then the only tests that failed are the ones introduced by @olaurino

**Comment by DougBurke** [comment]
I think we should keep that feature in (data is integrated but model is
not).

Doug

**Comment by DougBurke** [comment]
As a check for #708 , I compared the plot from `master` with the script in #708 which shows the wavy component:

master

to that with this branch (here the top plot is `plot_fit` and the bottom is `plot_data` + `plot_source_component` since the models now overlap if drawn on the same graph):

branch

**Comment by DougBurke** [comment]
@dtnguyen2 - in trying some of the tests I added in #724 I am seeing some odd behavior with this PR. The following code creates a box set to 0.4 for x>=3.1 and < 4.2 and zero outside, with the "data" grid being x=2,2.1,2.2,...,6. The "regrid" grid partially overlaps this (in this case 2.5,2.7,2.9,...7 but it happens with a range of different overlaps). When evaluating with this "regrid", only a small number of output bins are set to 0.4, and they aren't consecutive:

```python
import numpy as np

from sherpa.models.basic import Box1D
from sherpa.data import Data1D

xgrid = np.arange(2, 6, 0.1)
d = Data1D('tst', xgrid, np.ones_like(xgrid))

mdl = Box1D()
mdl.xlow = 3.1
mdl.xhi = 4.2
mdl.ampl = 0.4

yexpected = d.eval_model(mdl)
print("Expected: {} - {}".format(yexpected.min(), yexpected.max()))

requested = np.arange(2.5, 7, 0.2)
# requested = np.arange(1, 5.1, 0.075)

rmdl = mdl.regrid(requested)
ygot = d.eval_model(rmdl)

for (x, yg, ye) in zip(xgrid, ygot, yexpected):
    print(" x={:.2f}  expected={}  got={}".format(x, ye, yg))
```

produces the following output (I've edited in two lines to indicate the start/end of the box model):

```
In [1]: %run mdls.py
Expected: 0.0 - 0.4
 x=2.00  expected=0.0  got=0.0
 x=2.10  expected=0.0  got=0.0
 x=2.20  expected=0.0  got=0.0
 x=2.30  expected=0.0  got=0.0
 x=2.40  expected=0.0  got=0.0
 x=2.50  expected=0.0  got=0.0
 x=2.60  expected=0.0  got=0.0
 x=2.70  expected=0.0  got=0.0
 x=2.80  expected=0.0  got=0.0
 x=2.90  expected=0.0  got=5.91645678915757e-30
 x=3.00  expected=0.0  got=0.0
--------------------------------------------------note: expect 0 above, 0.4 below this
 x=3.10  expected=0.4  got=0.4
 x=3.20  expected=0.4  got=0.0
 x=3.30  expected=0.4  got=0.4
 x=3.40  expected=0.4  got=0.0
 x=3.50  expected=0.4  got=0.4
 x=3.60  expected=0.4  got=0.0
 x=3.70  expected=0.4  got=0.4
 x=3.80  expected=0.4  got=0.0
 x=3.90  expected=0.4  got=0.4
 x=4.00  expected=0.4  got=0.0
 x=4.10  expected=0.4  got=0.4
------------------------------------------------------- note expect 0.4 above this 0 below
 x=4.20  expected=0.0  got=0.0
 x=4.30  expected=0.0  got=0.0
 x=4.40  expected=0.0  got=0.0
 x=4.50  expected=0.0  got=0.0
 x=4.60  expected=0.0  got=0.0
 x=4.70  expected=0.0  got=0.0
 x=4.80  expected=0.0  got=0.0
 x=4.90  expected=0.0  got=0.0
 x=5.00  expected=0.0  got=0.0
 x=5.10  expected=0.0  got=0.0
 x=5.20  expected=0.0  got=0.0
 x=5.30  expected=0.0  got=0.0
 x=5.40  expected=0.0  got=0.0
 x=5.50  expected=0.0  got=0.0
 x=5.60  expected=0.0  got=0.0
 x=5.70  expected=0.0  got=0.0
 x=5.80  expected=0.0  got=0.0
 x=5.90  expected=0.0  got=0.0
```

**Comment by DougBurke** [comment]
#764 shows the "partial overlap" cases that fail (rebased onto this PR to make it easier to run)

**Comment by DougBurke** [comment]
Yay - 88bec2b has fixed my problem!

**Pull request merged**